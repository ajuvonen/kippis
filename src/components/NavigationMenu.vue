<script setup lang="ts">
import {useI18n} from 'vue-i18n';
import {RouterLink} from 'vue-router';

const {t} = useI18n();
</script>
<template>
  <ul
    class="fixed h-14 bottom-0 sm:top-0 w-full z-10 bg-white flex flex-nowrap justify-center items-stretch shadow-sm"
    role="navigation"
  >
    <RouterLink to="/">{{ t('menu.home') }}</RouterLink>
    <RouterLink to="/search">{{ t('menu.search') }}</RouterLink>
    <RouterLink to="/selection">{{ t('menu.selection') }}</RouterLink>
  </ul>
</template>
<style lang="scss" scoped>
a {
  @apply flex flex-col justify-center w-1/3 sm:w-28 text-center m-1 rounded-md uppercase text-sm tracking-wide transition-colors ease-in-out;

  &:hover {
    @apply bg-slate-800 text-white;
  }

  &.router-link-exact-active:not(:hover) {
    @apply bg-slate-100;
  }
}
</style>
