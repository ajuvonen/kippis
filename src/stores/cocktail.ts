import {
  getByIngredients,
  getBySearchString,
  getDetails,
  getNonAlcoholic,
  getByFirstLetter,
} from '@/api';
import {SEARCHABLE_ALCOHOLS} from '@/utils/constants';
import type {FullDetailsCocktail, SearchResultCocktail, SearchStringProps} from '@/utils/types';
import {defineStore} from 'pinia';
import {compose, flatten, map, prop, sortBy, uniq, without} from 'ramda';

export const useCocktailStore = defineStore('cocktail', {
  state: () => ({
    currentSearch: {} as SearchStringProps,
    cocktailDetails: [] as FullDetailsCocktail[],
    preventFetch: false,
    searchResults: [] as SearchResultCocktail[],
    selection: new Set<number>(),
    highlightedCocktail: null as number | null,
  }),
  getters: {
    getCocktailDetails: (state) => (id: number) => state.cocktailDetails.find((cocktail) => cocktail.id === id) || null,
    // Get a "shopping list" of ingredients. They should be unique, sorted and without some obvious items
    getAllIngredients(state): string[] {
      return compose(
        sortBy(prop(0)),
        without(['water', 'ice']),
        uniq,
        flatten<string[][]>,
        map((id: number) =>
          // Make cream, egg, and citrus fruits show up as their base form 
          this.getCocktailDetails(id)!.ingredients.map(({ingredient}) =>
            ingredient.replace(/^whipped | peel$| spiral$| white$| yolk$/, ''),
          ),
        ),
      )(Array.from(state.selection));
    },
  },
  actions: {
    async addToSelection(id: number) {
      await this.fetchCocktail(id);
      this.selection.add(id);
    },
    async fetchCocktail(id: number) {
      if (!this.cocktailDetails.some((cocktail) => cocktail.id === id)) {
        this.cocktailDetails.push(await getDetails(id));
      }
    },
    async openCocktailModal(id: number) {
      await this.fetchCocktail(id);
      this.highlightedCocktail = id;
    },
    removeFromSelection(id: number) {
      this.selection.delete(id);
    },
    async search(searchString: string) {
      if (searchString.length === 1) {
        this.searchResults = await getByFirstLetter(searchString);
      } else {
        this.searchResults = await getBySearchString(searchString);
      }
    },
    async searchNonAlcoholic() {
      this.searchResults = await getNonAlcoholic();
    },
    async searchWithTag(tag: string) {
      this.searchResults = await getByIngredients(
        SEARCHABLE_ALCOHOLS.find((alcohol) => tag === alcohol.tag)?.ingredients!,
      );
    },
  },
});
