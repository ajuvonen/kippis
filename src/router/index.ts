import {createRouter, createWebHistory} from 'vue-router';
import {isEmpty} from 'remeda';
import HomeView from '@/views/HomeView.vue';
import {useCocktailStore} from '@/stores/cocktail';

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'home',
      component: HomeView,
    },
    {
      path: '/search',
      name: 'search',
      props: (route) => ({
        searchString: !route.query.tag ? route.query.searchString : undefined,
        tag: route.query.tag,
      }),
      beforeEnter: (to) => {
        if (isEmpty(to.query)) {
          const cocktailStore = useCocktailStore();
          if (cocktailStore.currentSearch.tag || cocktailStore.currentSearch.searchString) {
            cocktailStore.preventFetch = true;
            return {path: to.path, query: {...cocktailStore.currentSearch}};
          }
        }
      },
      component: () => import('@/views/SearchView.vue'),
    },
    {
      path: '/ingredients',
      name: 'ingredients',
      component: () => import('@/views/IngredientsView.vue'),
    },
  ],
});

export default router;
