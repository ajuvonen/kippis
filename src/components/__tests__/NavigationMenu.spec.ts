import {mount} from '@vue/test-utils';
import {describe, it, expect, beforeEach} from 'vitest';
import router from '@/router';
import {useCocktailStore} from '@/stores/cocktail';
import NavigationMenu from '@/components/NavigationMenu.vue';
import {testCocktails} from '@/components/__tests__/mswHandlers';

describe('NavigationMenu', () => {
  const cocktailStore = useCocktailStore();

  beforeEach(() => {
    cocktailStore.$reset();
  });

  it('mounts', () => {
    const wrapper = mount(NavigationMenu);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows selection size', () => {
    cocktailStore.selection.push(testCocktails[0][1]);
    const wrapper = mount(NavigationMenu);
    expect(wrapper.findByTestId('selection-size-indicator').text()).toBe('1');
  });

  it.each(router.getRoutes())('includes link to $path', () => {
    const wrapper = mount(NavigationMenu);
    router.getRoutes().forEach((route) => {
      expect(wrapper.find(`[href="${route.path}"]`).exists()).toBe(true);
    });
  });
});
