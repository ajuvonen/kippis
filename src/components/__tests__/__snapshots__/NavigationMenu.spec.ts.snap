// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationMenu > mounts 1`] = `
"<nav data-v-fe35c617="" class="navigation-menu">
  <ul data-v-fe35c617="" class="navigation-menu__link-container">
    <li data-v-fe35c617=""><a data-v-fe35c617="" href="/" class="navigation-menu__link">Home</a></li>
    <li data-v-fe35c617=""><a data-v-fe35c617="" href="/search" class="navigation-menu__link">Search</a></li>
    <li data-v-fe35c617=""><a data-v-fe35c617="" href="/instructions" class="navigation-menu__link relative">Guide
        <!--v-if-->
      </a></li>
    <li data-v-fe35c617=""><a data-v-7747040a="" data-v-fe35c617="" href="#" role="button" class="navigation-menu__link navigation-menu__locale-changer gap-1" aria-label="Change language to fi"><svg data-v-7747040a="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
          <path d="M17.9,17.39C17.64,16.59 16.89,16 16,16H15V13A1,1 0 0,0 14,12H8V10H10A1,1 0 0,0 11,9V7H13A2,2 0 0,0 15,5V4.59C17.93,5.77 20,8.64 20,12C20,14.08 19.2,15.97 17.9,17.39M11,19.93C7.05,19.44 4,16.08 4,12C4,11.38 4.08,10.78 4.21,10.21L9,15V16A2,2 0 0,0 11,18M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z" class="transition-all"></path>
        </svg> en</a></li>
  </ul>
</nav>"
`;

exports[`NavigationMenu > shows selection size 1`] = `
"<nav data-v-fe35c617="" class="navigation-menu">
  <ul data-v-fe35c617="" class="navigation-menu__link-container">
    <li data-v-fe35c617=""><a data-v-fe35c617="" href="/" class="navigation-menu__link">Home</a></li>
    <li data-v-fe35c617=""><a data-v-fe35c617="" href="/search" class="navigation-menu__link">Search</a></li>
    <li data-v-fe35c617=""><a data-v-fe35c617="" href="/instructions" class="navigation-menu__link relative">Guide <div data-v-fe35c617="" class="navigation-menu__selection-size">1</div></a></li>
    <li data-v-fe35c617=""><a data-v-7747040a="" data-v-fe35c617="" href="#" role="button" class="navigation-menu__link navigation-menu__locale-changer gap-1" aria-label="Change language to fi"><svg data-v-7747040a="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
          <path d="M17.9,17.39C17.64,16.59 16.89,16 16,16H15V13A1,1 0 0,0 14,12H8V10H10A1,1 0 0,0 11,9V7H13A2,2 0 0,0 15,5V4.59C17.93,5.77 20,8.64 20,12C20,14.08 19.2,15.97 17.9,17.39M11,19.93C7.05,19.44 4,16.08 4,12C4,11.38 4.08,10.78 4.21,10.21L9,15V16A2,2 0 0,0 11,18M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z" class="transition-all"></path>
        </svg> en</a></li>
  </ul>
</nav>"
`;
